Feat/alert/model impl: Create AlertModelSupabase

agonzalez-r opened the pull request:
# Feat/alert/model-impl: Implementation of the Alert model with supabase

## Description
This PR introduces the class implementation (`AlertModelSupabase`) of the previously done model interface (`AlertModel`), using supabase as the database. 

I have some doubts about this implementation and the model in general, some of them are: 
- do we really need to declare the parameters `id` and `uid` in the data class `Alert`? or should these values be managed directly by supabase ? 
- should we add a function `getMyAlerts(uid, ...)`, giving the `uid` as parameter and selecting all alerts linked to that uid 

## Changes

## Files 

#### Added
- `main/.../model/alert/AlertModelSupabase`: implementation of all the functions defined in the interface 
#### Modified
- `main/.../model/alert/AlertModelSupabase`: had to add a parameter `idAlert` in the `updateAlert()` function to be able to identify which alert to modify 

## Testing
Not possible at the moment (View Model will be tested)

coaguila commented on app/src/main/java/com/android/periodpals/model/alert/AlertModelSupabase.kt:
```diff
@@ -0,0 +1,132 @@
+   */
+  override suspend fun addAlert(
+      alert: Alert,
+      onSuccess: () -> Unit,
```
you could return the alertId here

agonzalez-r replied:
That makes sense in the case where the idAlert is defined once it is stored in supabase but right now it is a part of the data class so the parameter `alert` that is passed to the function should already have an idAlert, possibly generated in the view model, right? 

I think it really depends on the implementation but it is one of the questions I asked in the PR description, did you take a look at them?

agonzalez-r replied:
or maybe we can turn the parameter nullable `val id: String? = ""` so that when the alert is created, we leave this value empty and the id is created directly in supabase, but still accessible through the data class?

coaguila commented on app/src/main/java/com/android/periodpals/model/alert/AlertModelSupabase.kt:
```diff
@@ -0,0 +1,132 @@
+    try {
+      val result =
+          supabase.postgrest[ALERTS].select { filter { eq("id", idAlert) } }.decodeSingle<Alert>()
+      onSuccess(result)
```
Don't forget to add `Log.d` before `onSuccess` to help debugging later on.

agonzalez-r commented:
# Feat/alert/model-impl: Implementation of the Alert model with supabase
## Description
This PR closes issue #117 by implementing and testing the functions declared in the interface `AlertsModel`. It also fixed some issues with the data class declaration, after further investigation about the proper methodology.

For better understanding of why some values in the `Alert` data class are nullable, check the table in supabase and their corresponding default values (some values are filled while the instance is created in supabase). 
## Changes
## Files 
#### Added
- `model/alert/AlertDto.kt`: data transfer object which is a serializable version of `Alert.kt`. This type of object is the one used for interacting with the repository, while the normal data class is used for UI usability. 

#### Modified
- `model/alert/Alert.kt`: made non-serializable, changed the type of some values (specially `createdAt` value was converted to string because of conflicts with "serializing" `LocalDateTime`)
- `model/alert/AlertModelSupabase`: ensured proper use of the `AlertDto` data class when calling repository functions and also did proper conversion to `Alert` when the functions had values to return. 
 
## Testing
- `test[UnitTest]/.../model/alert/AlertModelSupabaseTest`: tests for both success and failure (mocked) supabase responses to all the functions of the model.

charliemangano commented on app/src/main/java/com/android/periodpals/model/alert/AlertModelSupabase.kt:
```diff
@@ -0,0 +1,206 @@
+    try {
+      val result =
+          supabase.postgrest[ALERTS]
+              .select { filter { eq("id", idAlert) } }
```
What is the `eq` here? Does normal structural equality `==` not work?

charliemangano commented on app/src/main/java/com/android/periodpals/model/alert/AlertModelSupabase.kt:
```diff
@@ -0,0 +1,206 @@
+      Log.e(TAG, "getMyAlerts: fail to retrieve alerts: ${e.message}")
+      onFailure(e)
+    }
+  }
```
See `AlertModel` interface for my comment on this

charliemangano commented on app/src/main/java/com/android/periodpals/model/alert/AlertModelSupabase.kt:
```diff
@@ -0,0 +1,206 @@
+                createdAt = alert.createdAt,
+                location = alert.location,
+                message = alert.message,
+                status = alert.status)
```
Here too you can use the `AlertDto` constructor.

agonzalez-r replied:
That's right! forgot to use it

charliemangano commented on app/src/main/java/com/android/periodpals/model/alert/AlertModelSupabase.kt:
```diff
@@ -0,0 +1,206 @@
+          set("message", alertDto.message)
+          set("status", alertDto.status)
+        }) {
+          filter { eq("id", alert.id) }
```
Here as well, why the `eq`

agonzalez-r replied:
because we need to apply that only on that specific alert (with `id` == `alert.id`). Without the `eq` it applies that to all alerts in the table

charliemangano commented on app/src/main/java/com/android/periodpals/model/alert/AlertModelSupabase.kt:
```diff
@@ -0,0 +1,206 @@
+      onFailure: (Exception) -> Unit
+  ) {
+    try {
+      supabase.postgrest[ALERTS].delete { filter { eq("id", idAlert) } }
```
Again, `eq`?

agonzalez-r replied:
It's the only way to specify which alert.

charliemangano replied:
Oh yes you're right, I finally found the docs for it this morning

charliemangano commented on app/src/test/java/com/android/periodpals/model/alert/AlertModelSupabaseTest.kt:
```diff
@@ -0,0 +1,220 @@
+
+  private lateinit var alertModelSupabase: AlertModelSupabase
+
+  companion object {
```
Weird mix of snake case and camel case ●﹏●

agonzalez-r replied:
Yes I followed it to be consistent with the other model unit tests :( like `AuthenticationModelSupabaseTest`, I just realised it's not the case for `UserModelSupabaseTest`

charliemangano commented on app/src/test/java/com/android/periodpals/model/alert/AlertModelSupabaseTest.kt:
```diff
@@ -0,0 +1,220 @@
+import com.android.periodpals.model.alert.Alert
```
Very complete tests, and very clean code as well, good job!

charliemangano reviewed (changes requested):
**WIP sorry je vais dodo mais j'ai peur que ça s'enregistre pas mdr**

> do we really need to declare the parameters id and uid in the data class Alert? or should these values be managed directly by supabase ?

I'm leaving this question to @coaguila, he'll know much more about this.

> should we add a function getMyAlerts(uid, ...), giving the uid as parameter and selecting all alerts linked to that uid

No, this should be done in the view mode. The model should stay as agnostic as possible of what is happening in the rest of the app. However, we can have a `getAlertsFilteredBy` function taking a `(Alert) -> Boolean` lambda as parameter. This would let the view model handle the actual logic but would make the model much more usable while still staying agnostic.

good job updating the docs
make sure to delete comments and TODOs, or to explicit them. Comments should be about the why not the how, and TODOs should be converted to issues or tasks.

agonzalez-r commented on app/src/main/java/com/android/periodpals/model/alert/Alert.kt:
```diff
@@ -1,34 +1,29 @@
+    val status: Status
 )
 
 /** Enum class representing the product requested with the alert. */
```
Yes that's true!

agonzalez-r commented on app/src/main/java/com/android/periodpals/model/alert/AlertModel.kt:
```diff
@@ -31,10 +32,25 @@ interface AlertModel {
    * Updates an existing alert (edited).
    *
-   * @param alert The alert to be updated.
+   * @param alert Updated parameters for the Alert, id must not be null to know which alert to
```
I wanted to specify that it is the **alert with the updated parameters** that you need to pass as argument. And since to find which alert it is we need its `id`, we need that to be non-null inside the Alert. 
Could you propose a better way of describing this?

charliemangano replied:
I'd say something along the lines of:
```suggestion
   * @param alert Alert with updated parameters. `id` must not be null to know which alert to
```

agonzalez-r commented on app/src/main/java/com/android/periodpals/model/alert/AlertModel.kt:
```diff
@@ -31,10 +32,25 @@ interface AlertModel {
+      uid: String,
+      onSuccess: (List<Alert>) -> Unit,
+      onFailure: (Exception) -> Unit
+  )
```
Okay, I can change that !

coaguila reviewed (approved):
Lgtm

charliemangano commented on app/src/main/java/com/android/periodpals/model/alert/AlertModelSupabase.kt:
```diff
@@ -0,0 +1,186 @@
+   *   alerts as a parameter.
+   * @param onFailure Callback function to be called on failure, with the exception as a parameter.
+   */
+  override suspend fun getAlertsFilteredBy(
+      cond: PostgrestFilterBuilder.() -> Unit,
+      onSuccess: (List<Alert>) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    try {
+      val result = supabase.postgrest[ALERTS].select { filter(cond) }.decodeList<AlertDto>()
+      Log.d(TAG, "getMyAlerts: Success")
+      onSuccess(result.map { it.toAlert() })
+    } catch (e: Exception) {
+      Log.e(TAG, "getMyAlerts: fail to retrieve alerts: ${e.message}")
+      onFailure(e)
+    }
+  }
```
Good generalisation of old `getMyAlerts` into `getAlertsFilteredBy`, thanks for implementing my comment suggestion!

charliemangano reviewed (approved):
LGTM! Thanks for implementing the points I had highlighted.
